src/include/duckdb/main/database.hpp

```diff
@@ -99,8 +99,12 @@ class DuckDB {
 	DUCKDB_API FileSystem &GetFileSystem();
 
 	DUCKDB_API idx_t NumberOfThreads();
-	DUCKDB_API static const char *SourceID();
-	DUCKDB_API static const char *LibraryVersion();
+	__attribute__((__visibility__("default"))) static constexpr const char *SourceID() {
+		return DUCKDB_SOURCE_ID;
```

Comment:
I don't quite follow why this solves the problem - in both cases the version should be hard-coded in the library, no?

Reply:
My understanding is that the current version of LibraryVersion() when called (on any codepath) becomes a call to the linked DuckDB LibraryVersion, that if they are the same makes the check void (in the not currently used EXTENSION_STATIC_BUILD=0).

Reply:
I kind of wonder if this will work - this will require extensions to have `DUCKDB_SOURCE_ID` defined while compiling the extension itself.

Perhaps a nicer solution is to avoid having the extension contain the version function (`parquet_version`, etc) entirely and have it be generated in an additional .cpp file as part of `build_loadable_extension` - then the generated file can contain the hard-coded version constant.

Reply:
Right, so then this has gotten a bit out of hand, at first I did not saw that the defines had to be global, and this is not really cool (recompile all!), your suggestion make sense, but that or also other way out would require some changes to the interfaces that might make the problem less local and worth a generic look at it.

Reply:
We might even want to consider generating all extension C functions (`_init`, `_version` as well as `_storage_init`), but yes this would definitely be a bigger fix that requires changing all extensions

Reply:
CC @samansmink

Reply:
Thanks for the comments, moving this offline at some point in the near future, this can still be useful as a starting point.